### DISCUSSION TITLE
프로젝트 사용 기술

### DISCUSSION THREAD

**Discussion by jopopcorn** [euphony-io/check-in-speaker#3] [General]
@euphony-io/euphoria 

안드로이드 개발을 위해 필요한 기술, 아키텍처, 컨벤션에 대한 의견을 자유롭게 내주세요!
제가 생각해본 기술은 다음과 같습니다. 

언어: Kotlin
아키텍처: MVVM (AAC ViewModel)
라이브러리: ViewBinding / DataBinding, LiveData, ...
레이아웃: ConstraintLayout

</br>

- [x] 코드 컨벤션
- [x] 깃 컨벤션
- [x] 기능 추리기
- [x] 데드라인 정하기 (성호, 하은, 수연)
- [x] 기능별 역할 정하기 (성호, 수연)

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
다들 코틀린 사용해보셨나요? 가능하면 코틀린으로 하면 좋을 것 같습니다!
만약 코틀린을 사용해보지 않았더라도 자바를 알고 있다면 충분히 적용할 수 있습니다! 모르신다면 같이 스터디 해요 😸

**Reply by psh2849** [reply]
저도 코틀린으로 하는 게 더 익숙한 것 같습니다!
추가로, 데이터베이스를 사용한다면 어떤 걸 사용할지 의논하면 좋을 것 같습니다.
이전 샘플 앱 프로젝트에서는 Room을 사용했었는데 더 좋은 게 있다면 말씀해주세요! 
그리고.. 아키텍처 패턴은 따로 조금 더 공부해오도록 하겠습니다 ㅎㅎ

**Reply by jopopcorn** [reply]
데이터베이스를 사용한다면 유저 정보를 미리 넣어놓고 불러오기할 용도로 사용하는 건가요?
Room도 괜찮을 것 같고, 간단하게 데이터를 저장할 용도라면 SharedPreferences도 괜찮을 거 같아요!

아키텍처는 어렵게 생각하지 않으셔도 됩니다!
안드로이드에서의 MVVM 패턴은 보통 이야기하는 MVVM이랑 달라서 그 결을 이해하는 게 좋을 것 같아요.
아래 링크 참고하심 좋을 거 같고, 안드로이드 공식 문서의 ViewModel + 코드랩 따라해보시면 도움될 것 같아요!
https://wooooooak.github.io/android/2019/05/07/aac_viewmodel/

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 

상용 앱들을 확인해봤는데, 맨 처음 체크인 앱 화면을 실행하면 `사용자 정보 수집에 동의하는지 체크 + 전화번호 인증 화면`이 있더라고요
이것까지 추가로 하면 화면을 두개로 나누어도 괜찮을 것 같습니다! 혹은 추가로 `스플래쉬 화면`까지 만들어도 괜찮을 것 같아요!

필요한 젯팩 라이브러리도 생각해봤는데, 디자인한 화면을 토대로 개발한다면 액티비티 하나만 있으면 돼서 간단하게만 작업하면 될 것 같아요
`ViewModel+LiveData` 사용하고 레이아웃 결합은 `viewbinding`이랑 `databinding` 중 취향것 사용하면 될 것 같습니다!

필요한 기능도 생각해봤는데, 디자인한 메인 화면에서의 기능은 이정도로 추려질 것 같습니다!
- 음파 생성을 위한 마이크 권한 받기
- 다이얼로그 띄우고 전화번호 입력받기
- 전화번호 `SharedPreferences`에 저장하기 (안심번호로 변환 후 저장할 건지 고려)
- 안심번호로 변환하기 (변환 후 안심번호만 내부에 저장 or 전화번호와 안심번호 둘 다 저장)
- 체크인 버튼 눌렀을 때 음파 송신 시작/종료를 제어
- 음파 송신 중에는 빙글빙글 도는 프로그레스바 보이기

**Reply by psh2849** [reply]
정리 감사합니다!

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 

정규 회의 때 멘토님께 피드백 받은 대로 개발 규모를 심플하게 가는 게 좋을 것 같습니다!
따라서 화면은 디자인한 UI 대로 한 화면만 구현하면 될 것 같고, 전화번호를 안심번호로 변환하는 과정도 생략하면 될 것 같습니다!

피드백 + 지금까지 나온 개발 시나리오를 토대로 해야 할 것들을 다시 정리해보았습니다.

</br>

**UI**
- 안심번호를 입력 받는 다이얼로그 만들기
- 메인 화면 xml 작업하기

</br>

**개발**
- 음파 생성을 위한 마이크 권한 받기
- 다이얼로그 띄우고 안심번호 입력받기 (기존에 발급받은 안심번호를 입력받고, 유효하지 않은 문자인지 간단히 체크)
- 안심번호 `SharedPreferences`에 저장하기
- 체크인 버튼 눌렀을 때 음파 송신 시작/종료를 제어
  - 체크인 시작/종료에 따라 소리 크기 조절 (`AudioManager`)
  - 오디오 포커싱 (`AudioManager`)
- 음파 송신 중 프로그레스바 보이기

</br>

수정했으면 하는 부분이나 추가・삭제하면 좋을 기능이 있다면 자유롭게 의견 나눕시다!

**Reply by designe** [reply]
안심번호를 실제 자기 자신의 안심번호를 저장해 사용할 수 있으면 kipass에 적용되지 않더라도 독자적으로도 앱으로 런칭되면 사용할 수 있지 않을까 하네요!

이런 기능은 어떨까요?
사용자가 안심번호로 음파를 사용했던 장소의 위치 정보(GPS 등?)를 저장해 표시해준다면 (리스트 등으로 표시하고, 누르면 맵이 보인다던지?)
내가 어디에 들렀었는지 개인 기록 저장도 되고 간략한 메세지 같은 것도 나중에 남길 수 있고 하면 좋겠다는 생각이 드네요 ㅎㅎ

**Reply by psh2849** [reply]
체크인 송신 시에 보내는 데이터는 안심번호만  보내주면 되는거죠?

**Reply by jopopcorn** [reply]
> 체크인 송신 시에 보내는 데이터는 안심번호만 보내주면 되는거죠?

네넵 맞습니다!!

**Reply by jopopcorn** [reply]
> 안심번호를 실제 자기 자신의 안심번호를 저장해 사용할 수 있으면 kipass에 적용되지 않더라도 독자적으로도 앱으로 런칭되면 사용할 수 있지 않을까 하네요!
> 
> 이런 기능은 어떨까요?
> 사용자가 안심번호로 음파를 사용했던 장소의 위치 정보(GPS 등?)를 저장해 표시해준다면 (리스트 등으로 표시하고, 누르면 맵이 보인다던지?)
> 내가 어디에 들렀었는지 개인 기록 저장도 되고 간략한 메세지 같은 것도 나중에 남길 수 있고 하면 좋겠다는 생각이 드네요 ㅎㅎ

좋은 생각이네요! 😲
그런데 음파 데이터 수신 완료를 감지하는 로직이 필요할 것 같은데, 리스너 앱에서 체크인이 됐을 때를 스피커 앱에서 확인할 수 있는 방법이 있을까요?
제 생각에는 푸시 알림같은 기능이 있어야 스피커 앱에서 정확한 체크인 기록을 저장할 수 있을 것 같아서요.. 
우선은 음파를 생성할 때마다 위치를 저장하게끔 하는게 좋을까요? GPS 관련해서 저도 알아보고 추가 개발 고려해보겠습니다!

**Reply by designe** [reply]
푸시 알림 같은 기능이 있으면 좋긴 하겠네요! 리스너앱 등에서 ack 패킷같은걸 발생시키면 될거 같긴 한데 복잡도가 올라가긴 하니 차후 업데이트 요소로 생각해두면 좋을것 같네요!  :)

**Reply by jopopcorn** [reply]
> 푸시 알림 같은 기능이 있으면 좋긴 하겠네요! 리스너앱 등에서 ack 패킷같은걸 발생시키면 될거 같긴 한데 복잡도가 올라가긴 하니 차후 업데이트 요소로 생각해두면 좋을것 같네요! :)

알겠습니다! 좋은 아이디어 감사드려요 😸

**Comment by jopopcorn** [comment]
@psh2849 @HACore 
데드라인을 미리 정하고 개발을 본격적으로 시작하는 게 좋을 것 같아요. 제가 생각한 최종 마감 기간은 넉넉 잡아 10월 6일(수)입니다.
다들 스케줄이 어떤지 확인하고 데드라인과 역할을 정하는 게 좋을 것 같습니다!

**작업을 끝낼 수 있는 날짜를 코멘트 달아주세요!**
추가로 성호님께서는 위에서 확정된 개발 기능 중에서 맡아보고 싶은 파트가 있으면 말씀해주세요!

**Reply by jopopcorn** [reply]
저는 다음주 평일 내내 중요한 일정이 있어서 10월 2일부터 2-3일간 빡세게 작업할 것 같아요!
그래도 최대한 하루에 30분에서 한시간 정도는 개발하려고 하고, 일찍 마무리 지을 수 있다면 10월 3일(일)에 개발을 끝내는 게 좋지 않을까 싶네요!

**Reply by psh2849** [reply]
- 음파 생성을 위한 마이크 권한 받기
- 체크인 버튼 눌렀을 때 음파 송신 시작/종료를 제어
- 음파 송신 중 프로그레스바 보이기

이렇게 한번 해보고 싶습니다! 저도 10월 3일까지 하는 것을 목표로 해볼게요
추가로 위에서 멘토님이 말씀해주신 것도 고려해봐야할 것 같아요

**Reply by designe** [reply]
@psh2849 음파 송신 중 프로그레스바 보이기 좋네요! :)

**Reply by jopopcorn** [reply]
> * 음파 생성을 위한 마이크 권한 받기
> * 체크인 버튼 눌렀을 때 음파 송신 시작/종료를 제어
> * 음파 송신 중 프로그레스바 보이기
> 
> 이렇게 한번 해보고 싶습니다! 저도 10월 3일까지 하는 것을 목표로 해볼게요
> 추가로 위에서 멘토님이 말씀해주신 것도 고려해봐야할 것 같아요

넵넵 좋습니다! 그럼 저는 다이얼로그 만들고 다이얼로그에서 안심번호 입력 받고 저장하는 기능 개발하겠습니다!
그리고 위에서 멘토님께서 말씀주신 아이디어(위치 정보 저장 기능)은 따로 조사해보고 새로운 코멘트에 올리겠습니다
내일 화상회의 때도 이부분 추가로 이야기 나누어보면 좋을 것 같아요~!

**Reply by psh2849** [reply]
> > * 음파 생성을 위한 마이크 권한 받기
> > * 체크인 버튼 눌렀을 때 음파 송신 시작/종료를 제어
> > * 음파 송신 중 프로그레스바 보이기
> > 
> > 이렇게 한번 해보고 싶습니다! 저도 10월 3일까지 하는 것을 목표로 해볼게요
> > 추가로 위에서 멘토님이 말씀해주신 것도 고려해봐야할 것 같아요
> 
> 넵넵 좋습니다! 그럼 저는 다이얼로그 만들고 다이얼로그에서 안심번호 입력 받고 저장하는 기능 개발하겠습니다!
> 그리고 위에서 멘토님께서 말씀주신 아이디어(위치 정보 저장 기능)은 따로 조사해보고 새로운 코멘트에 올리겠습니다
> 내일 화상회의 때도 이부분 추가로 이야기 나누어보면 좋을 것 같아요~!

넵 저도 한번 어떻게 구현하면 좋을지 생각해보고 있을게요!

**Reply by HACore** [reply]
저는 화요일 중으로 메인 화면 xml 작업 완료해서 PR올릴게요!

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
코드 컨벤션은 저번주 정규 회의 때 멘토님이 설명해주신 [Kotlin 클린코드 네이밍 규칙](https://docs.google.com/presentation/d/1c_JZAMIV7uzTVJGehCpLAaCeDR-UjDr4/edit#slide=id.p4)을 따르면 좋을 것 같습니다!

그리고 리소스 네이밍 컨벤션은 아래 첨부한 이미지의 규칙을 따르는 편이라서 여기에 맞춰서 id를 지어주면 더 좋을 것 같아요
예를 들어서 텍스트뷰는 tv_user_name, 에딧텍스트는 et_phone_number 와 같이 이름을 지어줬습니다!
<img width="70%" alt="android resource convention" src="https://user-images.githubusercontent.com/35393459/113091935-f400b900-9227-11eb-9a1e-188ad085387c.png"/>
커밋 컨벤션은 저번 팀에서는 [이 규칙](https://blog.ull.im/engineering/2019/03/10/logs-on-git.html)을 참고해서 커밋 메시지를 작성했는데, 보완했으면 하는 부분 있다면 자유롭게 의견 남겨주세요!

**Reply by psh2849** [reply]
오 이미지 규칙 되게 좋네요..
tv, et, iv 이런거만 알았는데 저런 것도 도입하면 되게 쉽게 알아 볼 수 있을 것 같습니다 ㅎㅎ

**Reply by yeonns** [reply]
android resource naming 규칙도 있었네요! detail이 살아있는 naming 연구 멋져요 😆

**Reply by HACore** [reply]
오.. 주로 `<what>_<where>_<description>`으로 적는편인데 `<size>`가 포함된 버전도 있었군요..!
이번 마크업 작성 시에 해당 규칙 따라서 작성하겠습니다 👍👍

**Reply by kmsaidev** [reply]
감사합니다!! 잘 참고해서 개발하겠습니다 👍

**Comment by psh2849** [comment]
@euphony-io/euphoria 
Progress bar 관련해서 질문 드립니다!
Speak 할 때 전송이 끝났다는 걸 어떻게 알 수 있을까요?
Listener 쪽에는 notify 함수가 있었던 것 같은데 Speaker 쪽에는 따로 그런 함수가 없는 것 같아서요

**Reply by jopopcorn** [reply]
앗 그러네요 전송이 잘 마쳤을 때를 어떻게 확인하는지에 관해 멘토님과 상의한 [코멘트](https://github.com/euphony-io/wave-in-speaker/discussions/3#discussioncomment-1385666)가 이 페이지에 있어서 참고하면 좋을 것 같아요!
알림이나 ACK 패킷 보내기같은 기능을 당장 도입하기는 어려울 것 같은데요.
제 생각에는 체크인 시작/중단을 제어할 때 프로그레스바 visibility 처리만 해주는 게 좋을 것 같습니다 😿

**Reply by psh2849** [reply]
넵 알겠습니다. 그러면 visibility 처리만 일단 해주도록 할게요! 
그래도 사용자 입장에서는 Listener 기기에서 notify 됐을 때 전송 완료됐다는 Toast 메세지를 하나 출력해주면 쉽게 판단할 수 있을 것 같습니다!
